lib/autoproj/cli/ci.rb

```diff
@@ -50,22 +70,27 @@ def cache_pull(dir, ignore: [], silent: true)
                 results
             end
 
-            def cache_push(dir, force: [], silent: true)
+            def cache_push(dir, silent: true)
                 packages = resolve_packages
+                metadata = consolidated_report['packages']
 
-                built = load_built_flags
-
-                memo   = Hash.new
+                memo = {}
                 results = packages.each_with_object({}) do |pkg, h|
-                    unless built[pkg.name]
-                        next
+                    next unless (pkg_metadata = metadata[pkg.name])
+                    next unless (build_info = pkg_metadata['build'])
+                    next if build_info['cached'] || !build_info['success']
+
+                    # Remove cached flags before saving
+                    pkg_metadata = pkg_metadata.dup
+                    PHASES.each do |phase_name|
+                        pkg_metadata[phase_name]&.delete('cached')
                     end
 
                     state, fingerprint = push_package_to_cache(
-                        dir, pkg, force: force.include?(pkg.name), memo: memo)
+                        dir, pkg, pkg_metadata, force: true, memo: memo
+                    )
                     puts "pushed #{pkg.name} (#{fingerprint})" if state && !silent
```

Comment:
I will pretty soon have to output stuff on the daemon as well. My plan was to stick to Autoproj's reporting methods (info, error, warn). Is there a reason to use plain `puts` here?

Reply:
No good reason, for this one. I'll switch to `info`. Only MainCI#status should use puts as displaying is part of what MainCI#status is meant to do.